# -*- snakemake -*-
include: '../settings.rules'

DEFAULT_RULES = ['rsem_calculate_expression', 'rsem_prepare_reference']

config_default = { 
    'bio.ngs.rnaseq.rsem' : {
        'rules' : DEFAULT_RULES,
        'annot_gtf' : config['bio.ngs.settings']['annotation']['transcript_annot_gtf'],
        'threads' : config['settings']['threads'],
        'ref' : config['bio.ngs.settings']['db']['ref'],
        'ref_sfx' : '.transcripts.fa',
        'index' : "",
        'index_is_transcriptome': False, # Set to true if mapping directly to transcriptome
        'calculate-expression' : {
            'cmd' : 'rsem-calculate-expression',
            'options' : '--no-bam-output',
            'bowtie-options' : "--no-bam-output --bowtie-chunkmbs 512",
        },
    },
}

update_config(config_default, config)
config = config_default

# Add rsem to rnaseq section
config['bio.ngs.settings']['rnaseq']['_quantification'].append('rsem')
# Set transcriptome based quantification true (for STAR etc)
if not config['bio.ngs.rnaseq.rsem']['index_is_transcriptome']:
    config['bio.ngs.settings']['rnaseq']['_transcriptome_quantification'] = True

for rule in config['bio.ngs.rnaseq.rsem']['rules']:
    include: os.path.join("_rsem", rule + ".rule")
